src/main/kotlin/com/openlattice/postgres/PostgresResultSetAdapters.kt

```diff
@@ -26,71 +27,94 @@ internal class PostgresResultSetAdapters
+    }
+
+    return id to entity
 }
 
+/**
+ * Returns linked entity data from the [ResultSet] mapped respectively by its id, entity set, origin id and property
+ * type id.
+ * Note: Do not include the linking id for the [IdConstants.ID_ID] key as a property for this adapter, because it is
+ * used for linked entity indexing and we preserve that key for the origin id.
+ * @see ConductorElasticsearchImpl.formatLinkedEntity
+ */
 @Throws(SQLException::class)
-fun getEntityPropertiesByPropertyTypeId2(
+fun getEntityPropertiesByEntitySetIdOriginIdAndPropertyTypeId(
         rs: ResultSet,
         authorizedPropertyTypes: Map<UUID, Map<UUID, PropertyType>>,
+        metadataOptions: Set<MetadataOption>,
         byteBlobDataManager: ByteBlobDataManager
-): Pair<UUID, MutableMap<UUID, MutableSet<Any>>> {
+): Pair<UUID, Pair<UUID, Pair<UUID, MutableMap<UUID, MutableSet<Any>>>>> {
     val id = id(rs)
     val entitySetId = entitySetId(rs)
+    val originId = originId(rs)
     val propertyTypes = authorizedPropertyTypes.getValue(entitySetId)
-    val dataTypes = propertyTypes.map { (_, pt) -> PostgresEdmTypeConverter.map(pt.datatype) }.toSet()
-
-    val entity = dataTypes.map { datatype ->
-        val json = rs.getString("v_$datatype")
-        mapper.readValue<MutableMap<UUID, MutableSet<Any>>>(json)
-    }.fold(mutableMapOf(IdConstants.ID_ID.id to mutableSetOf<Any>(id))) { acc, mutableMap ->
-        acc.putAll(mutableMap)
-        return@fold acc
-    }
 
-    (entity.keys - propertyTypes.keys).forEach { entity.remove(it) }
-
-    propertyTypes.forEach { (_, propertyType) ->
+    val entity = readJsonDataColumns(
+            rs,
+            propertyTypes,
+            byteBlobDataManager
+    )
 
-        if (propertyType.datatype == EdmPrimitiveTypeKind.Binary) {
-            val urls = entity.getOrElse(propertyType.id) { mutableSetOf() }
-            if (urls.isNotEmpty()) {
-                entity[propertyType.id] = byteBlobDataManager.getObjects(urls).toMutableSet()
-            }
-        }
+    if (metadataOptions.contains(MetadataOption.LAST_WRITE)) {
+        entity[IdConstants.LAST_WRITE_ID.id] = mutableSetOf<Any>(lastWriteTyped(rs))
     }
 
-    return id to entity
+    return id to (entitySetId to (originId to entity))
 }
 
 
 @Throws(SQLException::class)
-fun getEntityPropertiesByPropertyTypeId3(
+fun getEntityPropertiesByFullQualifiedName(
         rs: ResultSet,
         authorizedPropertyTypes: Map<UUID, Map<UUID, PropertyType>>,
+        metadataOptions: Set<MetadataOption>,
         byteBlobDataManager: ByteBlobDataManager
 ): Pair<UUID, MutableMap<FullQualifiedName, MutableSet<Any>>> {
     val id = id(rs)
     val entitySetId = entitySetId(rs)
     val propertyTypes = authorizedPropertyTypes.getValue(entitySetId)
+
+    val entity = readJsonDataColumns(rs, propertyTypes, byteBlobDataManager)
+
+    val entityByFqn = entity.mapKeys { propertyTypes.getValue(it.key).type }.toMutableMap()
```

Comment:
Are we guaranteed that only authorized property types will be present in the results here?

Reply:
ohh I see -- readJsonDataColumns handles that

Reply:
Yep :)